File: src/main/scala/my/spark/examples/KBUserJoin.scala
package my.spark.examples

import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._

object KBUserJoin { 

  def getRequestDocId(s: String): String = { 
    val matchre = "KBDOC-[0-9]*".r
    matchre.findFirstIn(s).orNull
  }

  def main(args: Array[String]) {
  
    val kblist = "kblist.txt"
    val weblogs = "weblogs"

    val sc = new SparkContext()

    val docs =  sc.textFile(kblist).map(_.split(":")).map(tokens => (tokens(0),tokens(1))).cache()
    println("\n*** Document Id & Title")
    docs.take(5).foreach(println)

    val logs = sc.textFile(weblogs).map(line => (getRequestDocId(line),line.split(" ")(2))).filter(pair => pair._1 != null).distinct().cache()
    println("\n*** DocId and UserId")
    logs.take(5).foreach(println)

    val userDocs = logs.join(docs).map{case (docId,(userId,title)) => (userId,title)}.groupByKey()
    println("\n*** UserId and List of titles")
    for ((userId,titles) <- userDocs.take(3)) {
      println("userid: " + userId)
      for (title <- titles) println ("\t" + title)
    }
    println()
    sc.stop()
  }
}

File: src/main/python/KBUserJoin.py
import sys
import re
from pyspark import SparkContext

if __name__ == "__main__":
  
  kblist = "kblist.txt"
  weblogs = "weblogs"

  sc = SparkContext()
  
  def getRequestDoc(s):
    match=re.search(r'KBDOC-[0-9]*',s)
    return match.group() if match else None

  docs = sc.textFile(kblist).map(lambda line: line.split(':')).map(lambda tokens:(tokens[0],tokens[1])).cache()
  print("\n*** Document Id & Title")
  for doc in docs.take(5): print '({0}, {1})'.format(doc[0], doc[1])

  logs = sc.textFile(weblogs).map(lambda line:(getRequestDoc(line),line.split(' ')[2])).filter(lambda (docId,userId): docId is not None).distinct()
  print("\n*** DocId and UserId")
  for log in logs.take(5): print '({0}, {1})'.format(log[0], log[1])

  userDocs = logs.join(docs).map(lambda (docId,(userId,title)):(userId,title)).groupByKey()
  print("\n*** UserId and List of titles")
  for (userId,titles) in userDocs.take(3):
    print "userid: ", userId
    for title in titles: print '\t',title
  print
  sc.stop()